Graphite: Backend functions endpoint

aangelisc opened the pull request:
This PR replicates the current frontend functions request to the backend. The backend request will only be used if the feature toggle `graphiteBackendMode` is enabled. With the feature toggle disabled the existing request will be used.

I've also done a minor refactor to support both `GET` and `POST` requests in a generic way. Additionally, I've refactored the `doGraphiteRequest` and `parseResponse` methods to take a parameter that supports returning the raw response, allowing handler functions to implement more specific parsing if needed.

I've updated the tests accordingly and expanded some of the test cases.

The frontend data source has had the required types added.

Part of grafana/data-sources#607

Fixes grafana/grafana#109822

Merge after #110678

aangelisc committed:
Add lint rules

aangelisc committed:
Backend decoupling

- Add standalone files
- Add graphite query type
- Add logger to Service
- Create logger in the ProvideService method
- Use a pointer for the HTTP client provider
- Update logger usage everywhere
- Update tracer type
- Replace simplejson with json
- Add dummy CallResource and CheckHealth methods
- Update tests

aangelisc committed:
Update ConfigEditor imports

aangelisc committed:
Update types imports

aangelisc committed:
Update datasource

- Switch to using semver package
- Update imports

aangelisc committed:
Update store imports

aangelisc committed:
Update helper imports and notification creation

aangelisc committed:
Update context import

aangelisc committed:
Update version numbers and logic

aangelisc committed:
Copy array_move from core

aangelisc committed:
Test updates

aangelisc committed:
Add required files and update plugin.json

aangelisc committed:
Update core references and packages

aangelisc committed:
Remove commented code

aangelisc committed:
Update wire

aangelisc committed:
Lint

aangelisc committed:
Fix import

aangelisc committed:
Copy null type

aangelisc committed:
Merge branch 'main' into andreas/decouple-graphite

aangelisc committed:
More lint

aangelisc committed:
Update snapshot

aangelisc committed:
Merge branch 'main' into andreas/decouple-graphite

aangelisc committed:
Merge branch 'main' into andreas/decouple-graphite

aangelisc committed:
Refactor backend

- Split query logic into separate file
- Move utils to separate file

aangelisc committed:
Add health-check logic

- Support backend healthcheck if the FF is enabled

aangelisc committed:
Remove query import support as unneeded

aangelisc committed:
Add test

aangelisc committed:
Add util function for decoding responses

aangelisc committed:
Add events types

aangelisc committed:
Add resource handler

aangelisc committed:
Add events handler and generic resource req handler

aangelisc committed:
Tests

aangelisc committed:
Update frontend

- Add types
- Update events function to support backend requests

aangelisc committed:
Lint and typing

aangelisc committed:
Lint

aangelisc committed:
Add metrics find endpoint

- Add types
- Add generic response parser
- Add endpoint
- Tests

aangelisc committed:
Update FE functoin to use backend endpoint

aangelisc committed:
Lint

aangelisc committed:
Simplify request

aangelisc committed:
Update test

aangelisc committed:
Metrics expand type

aangelisc committed:
Extract shared logic and add metric expand endpoint

aangelisc committed:
Update tests

aangelisc committed:
Call metric expand from backend

aangelisc committed:
Rename type for clarity

aangelisc committed:
Add get resource req handler

aangelisc committed:
Refactor doGraphiteRequest, parseResponse

Update tests

aangelisc committed:
Migrate functions endpoint to backend

aangelisc committed:
Add tests

aangelisc committed:
Review

aangelisc committed:
Merge branch 'main' into andreas/decouple-graphite

aangelisc committed:
Review

aangelisc committed:
Fix packages

aangelisc committed:
Merge branch 'andreas/decouple-graphite' into andreas/graphite-healthcheck

aangelisc committed:
Merge branch 'andreas/graphite-healthcheck' into andreas/graphite-events

# Conflicts:
#	pkg/tsdb/graphite/utils.go

aangelisc committed:
Merge branch 'andreas/graphite-events' into andreas/graphite-metrics-find

aangelisc committed:
Merge branch 'andreas/graphite-metrics-find' into andreas/graphite-metrics-expand

# Conflicts:
#	pkg/tsdb/graphite/resource_handler_test.go

aangelisc committed:
Merge branch 'andreas/graphite-metrics-expand' into andreas/graphite-functions

aangelisc committed:
Format

aangelisc committed:
Merge branch 'andreas/graphite-metrics-find' into andreas/graphite-metrics-expand

# Conflicts:
#	pkg/tsdb/graphite/resource_handler_test.go

aangelisc committed:
Merge branch 'andreas/graphite-metrics-expand' into andreas/graphite-functions

aangelisc committed:
Merge branch 'main' into andreas/graphite-healthcheck

# Conflicts:
#	pkg/tsdb/graphite/graphite.go
#	pkg/tsdb/graphite/query_test.go
#	public/app/plugins/datasource/graphite/datasource.ts

aangelisc committed:
Fix merge issues

aangelisc committed:
Merge branch 'andreas/graphite-healthcheck' into andreas/graphite-events

aangelisc committed:
Merge branch 'main' into andreas/graphite-events

# Conflicts:
#	pkg/tsdb/graphite/graphite.go
#	pkg/tsdb/graphite/utils.go

aangelisc committed:
Merge branch 'andreas/graphite-events' into andreas/graphite-metrics-find

aangelisc committed:
Merge branch 'andreas/graphite-metrics-find' into andreas/graphite-metrics-expand

aangelisc committed:
Merge branch 'andreas/graphite-metrics-expand' into andreas/graphite-functions

bossinc reviewed (approved):
minor nit but :tada:

aangelisc committed:
Merge branch 'main' into andreas/graphite-metrics-expand

# Conflicts:
#	pkg/tsdb/graphite/resource_handler.go
#	pkg/tsdb/graphite/resource_handler_test.go
#	pkg/tsdb/graphite/types.go

aangelisc committed:
Review

aangelisc committed:
Fix undefined values

aangelisc committed:
Extract request creation

- Add method for create requests generically with tests
- Replace usage in query method
- Update usages in resource handlers
- Update tests
- Update types

aangelisc committed:
Merge branch 'main' into andreas/graphite-metrics-expand

aangelisc committed:
Merge branch 'andreas/graphite-metrics-expand' into andreas/graphite-functions

# Conflicts:
#	pkg/tsdb/graphite/resource_handler.go
#	pkg/tsdb/graphite/resource_handler_test.go

aangelisc committed:
Lint

aangelisc committed:
Merge branch 'main' into andreas/graphite-functions

# Conflicts:
#	pkg/tsdb/graphite/resource_handler.go
#	pkg/tsdb/graphite/resource_handler_test.go

aangelisc committed:
Lint
